task_manager/task_manager.hh

```diff
@@ -0,0 +1,229 @@
+
+        virtual is_abortable is_abortable() const {
+            return is_abortable::no;
+        }
+
+        void set_number(int number) {
+            _status.number = number;
+        }
+
+        void set_type(std::string type) {
+            _status.type = type;
+        }
+
+        void change_state(task_state state) {
+            _status.state = state;
+        }
+
+        void set_keyspace(std::string keyspace) {
+            _status.keyspace = keyspace;
+        }
+
+        void set_table(std::string table) {
+            _status.table = table;
+        }
+
+        void set_entity(std::string entity) {
+            _status.entity = entity;
+        }
+
+        void set_total_progress(double progress) {
+            _status.progress.total = progress;
+        }
+
+        void report_completion_progress(double progress) {
+            _status.progress.completed += progress;
+        }
+
+        void start(task_manager& tm) {
+            _status.start_time = db_clock::now();
+            _status.state = task_manager::task_state::running;
+
+            (void)with_gate(tm.async_gate(), [f = done(), &tm, id = _status.id] () mutable {
+                return std::move(f).finally([&tm, id] {
+                    return sleep_abortable(tm.task_ttl(), tm.abort_source());
+                }).finally([&tm, id] {
+                    tm.unregister_task(id);
+                });
+            });
+        }
+
+        virtual task_ptr create_child(task_id uuid, foreign_task_ptr&& parent) const = 0;
+
+        future<> start_on_shards(task_manager& tm, std::vector<task_id>& uuids, task_ptr this_task) {
+            return tm.container().map([parent = seastar::make_foreign(this_task), &uuids] (task_manager& tm) mutable {
+                task_ptr task = parent->create_child(uuids[this_shard_id()], std::move(parent));
+                tm.register_task(task);
+                return seastar::make_foreign<task_ptr>(task);
+            }).then([this, &tm] (foreign_task_vector tasks) {
+                this->_children = std::move(tasks);
+                start(tm);
+            });
+        }
+
+        future<> done() {
+            return _done.get_shared_future();
+        }
+
+        void finish() {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::done;
+            _done.set_value();
+        }
+
+        void finish_failed(std::exception&& ex) {
+            _status.end_time = db_clock::now();
+            _status.state = task_manager::task_state::failed;
+            _status.error = ex.what();
+            _done.set_exception(ex);
+        }
+
+        virtual void abort() = 0;
+
+        friend bool operator==(const task_manager::task_ptr &lhs, const task_manager::task_ptr &rhs) {
+            return lhs->get_status().id == rhs->get_status().id;
+        }
+    };
+public:
+    task_manager(const db::config& cfg, abort_source& as) : _cfg(cfg), _as(as) {}
```

Comment:
1. noexcept
2. We don't want to carry the dependency on db::config too deep into the stack.  See `compaction_manager::config` for example, where we keep just the updateable_value, not the whole config. Then in main, we use a `sharded_parameter` to get the per-shard updateable value from db::config.